src/utils/temporal.cpp

```diff
@@ -486,21 +491,50 @@ std::pair<DateParameters, LocalTimeParameters> ParseLocalDateTimeParameters(std:
   }
 }
 
-LocalDateTime::LocalDateTime(const int64_t microseconds) {
-  auto chrono_microseconds = std::chrono::microseconds(microseconds);
-  static constexpr int64_t one_day_in_microseconds = std::chrono::microseconds{std::chrono::days{1}}.count();
-  if (microseconds < 0 && (microseconds % one_day_in_microseconds != 0)) {
-    date = Date(microseconds - one_day_in_microseconds);
+LocalDateTime::LocalDateTime(const int64_t microseconds)
+    : us_since_epoch_(std::chrono::sys_time<std::chrono::microseconds>(std::chrono::microseconds(microseconds))) {
+  // Already UTC
+}
+
+LocalDateTime::LocalDateTime(const DateParameters &date_parameters, const LocalTimeParameters &local_time_parameters) {
+  auto us_since_epoch_local = std::chrono::microseconds{Date{date_parameters}.MicrosecondsSinceEpoch() +
+                                                        LocalTime{local_time_parameters}.MicrosecondsSinceEpoch()};
+  const auto *tz = flags::run_time::GetTimezone();
+  if (tz) {
+    // APPLY TIMEZONE (local to UTC)
+    us_since_epoch_ =
+        tz->to_sys(std::chrono::local_time<std::chrono::microseconds>(std::chrono::microseconds(us_since_epoch_local)));
+  } else {
+    // Fallback to UTC
+    us_since_epoch_ = std::chrono::sys_time<std::chrono::microseconds>(std::chrono::microseconds(us_since_epoch_local));
+  }
+}
+
+LocalDateTime::LocalDateTime(const Date &date, const LocalTime &local_time) {
+  auto us_since_epoch_local =
+      std::chrono::microseconds{date.MicrosecondsSinceEpoch() + local_time.MicrosecondsSinceEpoch()};
+  const auto *tz = flags::run_time::GetTimezone();
+  if (tz) {
+    // APPLY TIMEZONE (local to UTC)
+    us_since_epoch_ =
+        tz->to_sys(std::chrono::local_time<std::chrono::microseconds>(std::chrono::microseconds(us_since_epoch_local)));
   } else {
-    date = Date(microseconds);
+    // Fallback to UTC
+    us_since_epoch_ = std::chrono::sys_time<std::chrono::microseconds>(std::chrono::microseconds(us_since_epoch_local));
   }
-  chrono_microseconds -= std::chrono::microseconds{date.MicrosecondsSinceEpoch()};
-  local_time = LocalTime(chrono_microseconds.count());
 }
 
-// return microseconds normilized with regard to epoch time point
+// return microseconds normalized with regard to epoch time point
+int64_t LocalDateTime::SysMicrosecondsSinceEpoch() const { return us_since_epoch_.time_since_epoch().count(); }
+
```

Comment:
I am not sure if I got things completely correct, but SysMicrosecondsSinceEpoch seems like the function used not to add offset one more time to `us_since_epoch_`. But since we correctly store `us_since_epoch_` in ctor, can we just have one function which returns MicroSecondsSinceEpoch without conversions?

Reply:
Awesome, got it now